### PR TITLE
Custom registry

### LINKED ISSUE #7: Use Promethereus pkg to expose metrics
# Problem

Currently `collector.collect` returns a string, representing the output for requests at  `/metrics`.
To be sure the output is always compatible with Prometheus, we should use its libraries to format that output.

# Implementation

Investigate Prometheus package and other examples like https://github.com/prometheus/node_exporter.

### PR DESCRIPTION
**Pull request by gravityblast**
This PR change the way we create the output sent from the metrics server using Prometheus structures and interfaces.

Description at [Issue #7](https://github.com/status-im/geth_exporter/issues/7) and initial discussion at https://github.com/status-im/status-go/pull/966

Usually, a Prometheus exporter has many collectors; each collector has many metric families; each family has one or many metrics.

[prometheus/node_exporter](https://github.com/prometheus/node_exporter) for example has:

```
* node_exporter (exporter)
  *  filesystem (collector)
    * device_error (family)
      * device_error{devive="x"} (metric)
      * device_error{devive="y"} (metric)
```

In that case, the metrics are created in the process with something like `prometheus.GaugeValue` and the server converts them in families and metrics depending on the type of metric.

In our case, we have only one collector for all the raw metrics collected calling `debug.metrics` via `IPC`. The type of each metric is unknown unless we hardcode them in this package. Currently, we allow all the metrics exported from `geth`, so if we add more metrics in `statusd`, they are automatically exported from `geth_exporter`.

We can change the default behaviour filtering metrics matching `overall`, or we can delegate the filtering part to the HTTP call, using the `collect[]` params (e.g. `?collect[overall]&collect[whisper]`).

Closes #7

### CODE REVIEW

File: collector.go
Comment (adriacidre) [inline collector.go]: Public method documentation

File: collector.go
Comment (adriacidre) [inline collector.go]: why `dto`? what does it mean?

File: handlers.go
Comment (adriacidre) [inline handlers.go]: I got confused when i was reading this `*ipcPath` variable.

Personally I think it would be clearer treating command line flags on `main.go` or naming it to something like `ipcPathFlag` so you have clear is something defined by the user.

File: collector.go
Comment (adriacidre) [inline collector.go]: Due to the new logic on this method you no longer need this conditional
```
if len(c.fs) == 0 {
    return true
}
```

File: collector.go
Comment (adriacidre) [inline collector.go]: I would rename it to `matchAll` or `matchAllFilters` or at least add a comment describing it will only return true if it matches all defined filters.

**Review by themue — COMMENTED** [review]
Looks good so far, but missing some godoc conforming comments for later maintainers. 😉

File: main.go
Comment (adriacidre) [inline main.go]: joining on a single line will make it read better
```
if err := http.ListenAndServe(listenAddress, nil); err != nil {
... 
}
```

File: registry.go
Comment (adriacidre) [inline registry.go]: should we return here?

File: registry.go
Comment (adriacidre) [inline registry.go]: Consider using named result parameters on this method declaration + naked results

**Review by adriacidre — COMMENTED** [review]
Looks good so far, added some comments :)

File: collector.go
Comment (gravityblast) [reply collector.go]: oh thank you, I missed this one in the refactoring, we can remove it, it's now in `registry`

File: collector.go
Comment (gravityblast) [reply collector.go]: data transfer object. I used the same names used in the official `node_exporter`

File: handlers.go
Comment (gravityblast) [reply handlers.go]: good idea,  I'm renaming all the flags and passing the `*ipcPathFlag` as a parameter to create the handler.

File: registry.go
Comment (gravityblast) [reply registry.go]: I followed the `Gatherer` interface docs that say that the response can be inclomplete in case of errors

https://github.com/prometheus/client_golang/blob/master/prometheus/registry.go#L126

File: collector.go
Comment (gravityblast) [reply collector.go]: yes I removed it, thank you!

**Comment by gravityblast** [comment]
@themue I added an issue to add all the docs https://github.com/status-im/geth_exporter/issues/12

**Comment by gravityblast** [comment]
@adriacidre I updated the code after your comments

**Review by adriacidre — APPROVED** [review]
Good job mate!

**Review by jakubgs — APPROVED** [review]
I'm new to Go but for what it's worth this looks good to me. Thanks for doing this work, I'm gonna use this in [infra-eth-cluster](https://github.com/status-im/infra-eth-cluster) as soon as you merge this.

**Review by adambabik — APPROVED** [review]
LGTM overall

The only thing that we could improve is getting rid of precalculated metrics like averages:
```
# HELP geth_whisper_envelopeSize_avgRate01Min metric exported from geth with debug.metrics
# TYPE geth_whisper_envelopeSize_avgRate01Min untyped
geth_whisper_envelopeSize_avgRate01Min 0
# HELP geth_whisper_envelopeSize_avgRate05Min metric exported from geth with debug.metrics
# TYPE geth_whisper_envelopeSize_avgRate05Min untyped
geth_whisper_envelopeSize_avgRate05Min 0
# HELP geth_whisper_envelopeSize_avgRate15Min metric exported from geth with debug.metrics
# TYPE geth_whisper_envelopeSize_avgRate15Min untyped
geth_whisper_envelopeSize_avgRate15Min 0
# HELP geth_whisper_envelopeSize_meanRate metric exported from geth with debug.metrics
# TYPE geth_whisper_envelopeSize_meanRate untyped
geth_whisper_envelopeSize_meanRate 0
# HELP geth_whisper_envelopeSize_overall metric exported from geth with debug.metrics
# TYPE geth_whisper_envelopeSize_overall untyped
geth_whisper_envelopeSize_overall 0
```

The only thing we should care here is `geth_whisper_envelopeSize_overall 0`.

Here is the method that produces the JSON output of metrics: https://github.com/status-im/status-go/blob/develop/vendor/github.com/ethereum/go-ethereum/node/api.go#L281-L392

We can pick only properties named: `Overall` and `Value`. 

The only tricky part is `metrics.Timer` where percentiles could be stored as a summary with quantiles. We can skip this though and create another issue as it's a more advanced thing.

**Pull request merged**
